#include "MockAgvClient.h"
#include <iostream>

using namespace lsk_muduo;
using namespace agv::proto;
using namespace agv::codec;

// ==================== æž„é€ ä¸Žç”Ÿå‘½å‘¨æœŸ ====================

MockAgvClient::MockAgvClient(EventLoop* loop,
                             const InetAddress& server_addr,
                             const std::string& agv_id,
                             double telemetry_freq,
                             double initial_battery,
                             double watchdog_timeout)
    : loop_(loop),
      client_(loop, server_addr, "MockAGV-" + agv_id),
      agv_id_(agv_id),
      connected_(false),
      state_(IDLE),
      battery_(initial_battery),
      x_(0.0),
      y_(0.0),
      theta_(0.0),
      telemetry_freq_(telemetry_freq),
      telemetry_interval_(1.0 / telemetry_freq),
      last_server_msg_time_(Timestamp::now()),
      watchdog_timeout_sec_(watchdog_timeout) {
    
    // è®¾ç½® TcpClient å›žè°ƒ
    client_.setConnectionCallback(
        std::bind(&MockAgvClient::onConnection, this, std::placeholders::_1));
    client_.setMessageCallback(
        std::bind(&MockAgvClient::onMessage, this,
                  std::placeholders::_1, std::placeholders::_2, std::placeholders::_3));
    
    LOG_INFO << "[MockAGV-" << agv_id_ << "] Created (freq=" << telemetry_freq 
             << "Hz, watchdog_timeout=" << watchdog_timeout << "s)";
}

MockAgvClient::~MockAgvClient() {
    LOG_INFO << "[MockAGV-" << agv_id_ << "] Destroyed";
}

// ==================== è¿žæŽ¥ç®¡ç† ====================

void MockAgvClient::connect() {
    LOG_INFO << "[MockAGV-" << agv_id_ << "] Connecting to server...";
    client_.connect();
}

void MockAgvClient::disconnect() {
    LOG_INFO << "[MockAGV-" << agv_id_ << "] Disconnecting...";
    client_.disconnect();
}

// ==================== TcpClient å›žè°ƒå®žçŽ° ====================

void MockAgvClient::onConnection(const TcpConnectionPtr& conn) {
    if (conn->connected()) {
        connected_ = true;
        conn_ = conn;
        LOG_INFO << "[MockAGV-" << agv_id_ << "] âœ“ Connected to server: "
                 << conn->peerAddress().toIpPort();
        
        // åˆ·æ–°æœåŠ¡å™¨æ¶ˆæ¯æ—¶é—´æˆ³ï¼ˆè¿žæŽ¥æˆåŠŸè§†ä¸ºæ”¶åˆ°æ¶ˆæ¯ï¼‰
        refreshServerMessageTime();
        
        // å¯åŠ¨å®šæ—¶å™¨
        loop_->runEvery(telemetry_interval_, 
                        std::bind(&MockAgvClient::onTelemetryTimer, this));
        loop_->runEvery(kHeartbeatIntervalSec, 
                        std::bind(&MockAgvClient::onHeartbeatTimer, this));
        loop_->runEvery(kBatteryUpdateIntervalSec, 
                        std::bind(&MockAgvClient::onBatteryTimer, this));
        loop_->runEvery(kWatchdogCheckIntervalSec, 
                        std::bind(&MockAgvClient::onWatchdogTimer, this));
        
        LOG_INFO << "[MockAGV-" << agv_id_ << "] Timers started";
        
    } else {
        connected_ = false;
        LOG_WARN << "[MockAGV-" << agv_id_ << "] âœ— Disconnected from server";
        
        // è§¦å‘ç´§æ€¥åœæ­¢ï¼ˆè¿žæŽ¥æ–­å¼€ï¼‰
        if (state_ != E_STOP) {
            setState(E_STOP);
            LOG_ERROR << "[EMERGENCY] Server connection lost!";
        }
    }
}

void MockAgvClient::onMessage(const TcpConnectionPtr& conn,
                              Buffer* buf,
                              Timestamp receive_time) {
    (void)conn;
    (void)receive_time;
    
    // åˆ·æ–°çœ‹é—¨ç‹—ï¼ˆæ”¶åˆ°ä»»ä½•æ¶ˆæ¯ï¼‰
    refreshServerMessageTime();
    
    // å¾ªçŽ¯å¤„ç†ç²˜åŒ…ï¼šä¸€ä¸ª Buffer å¯èƒ½åŒ…å«å¤šä¸ªå®Œæ•´æ¶ˆæ¯
    while (LengthHeaderCodec::hasCompleteMessage(buf)) {
        uint16_t msg_type = 0;
        uint16_t flags = 0;
        std::string payload;
        
        // ä½¿ç”¨ LengthHeaderCodec è§£ç 
        if (!LengthHeaderCodec::decode(buf, &msg_type, &payload, &flags)) {
            LOG_ERROR << "[MockAGV-" << agv_id_ << "] Failed to decode message";
            return;
        }
        
        // å¤„ç† Protobuf æ¶ˆæ¯
        handleProtobufMessage(msg_type, payload.data(), payload.size());
    }
}

// ==================== æ¶ˆæ¯å¤„ç†å®žçŽ° ====================

void MockAgvClient::handleProtobufMessage(uint16_t msg_type, 
                                          const char* payload, 
                                          size_t len) {
    switch (msg_type) {
        case MSG_AGV_COMMAND: {
            AgvCommand cmd;
            if (!cmd.ParseFromArray(payload, static_cast<int>(len))) {
                LOG_ERROR << "[MockAGV-" << agv_id_ << "] Failed to parse AgvCommand";
                return;
            }
            handleAgvCommand(cmd);
            break;
        }
        case MSG_HEARTBEAT: {
            Heartbeat msg;
            if (!msg.ParseFromArray(payload, static_cast<int>(len))) {
                LOG_ERROR << "[MockAGV-" << agv_id_ << "] Failed to parse Heartbeat";
                return;
            }
            handleHeartbeat(msg);
            break;
        }
        case MSG_NAVIGATION_TASK: {  // ã€è¿­ä»£ä¸‰æ–°å¢žã€‘
            NavigationTask task;
            if (!task.ParseFromArray(payload, static_cast<int>(len))) {
                LOG_ERROR << "[MockAGV-" << agv_id_ << "] Failed to parse NavigationTask";
                return;
            }
            handleNavigationTask(task);
            break;
        }
        default:
            LOG_WARN << "[MockAGV-" << agv_id_ << "] Unknown message type: 0x" 
                     << std::hex << msg_type << std::dec;
            break;
    }
}

void MockAgvClient::handleAgvCommand(const AgvCommand& cmd) {
    LOG_INFO << "[MockAGV-" << agv_id_ << "] [RECV] AgvCommand: cmd_type=" 
             << cmd.cmd_type() << " (" << CommandType_Name(cmd.cmd_type()) << ")";
    
    switch (cmd.cmd_type()) {
        case CMD_EMERGENCY_STOP:
            LOG_WARN << "[MockAGV-" << agv_id_ << "] Receiving EMERGENCY_STOP command";
            setState(E_STOP);
            break;
            
        case CMD_RESUME:
            LOG_INFO << "[MockAGV-" << agv_id_ << "] Receiving RESUME command";
            if (state_ == E_STOP || state_ == CHARGING) {
                setState(IDLE);
            }
            break;
            
        case CMD_PAUSE:
            LOG_INFO << "[MockAGV-" << agv_id_ << "] Receiving PAUSE command";
            if (state_ == MOVING || state_ == MOVING_TO_CHARGER) {
                setState(IDLE);
            }
            break;
            
        case CMD_REBOOT:
            LOG_WARN << "[MockAGV-" << agv_id_ << "] Receiving REBOOT command (ignored)";
            break;
            
        default:
            LOG_WARN << "[MockAGV-" << agv_id_ << "] Unknown command type: " 
                     << cmd.cmd_type();
            break;
    }
    
    // ç®€åŒ–å®žçŽ°ï¼šå°†ä»»ä½•æŒ‡ä»¤éƒ½è§†ä¸ºå……ç”µè§¦å‘ï¼ˆè¿­ä»£ä¸€ç®€åŒ–é€»è¾‘ï¼‰
    // æ³¨æ„ï¼šæŒ‰ç…§ GatewayServer.ccï¼Œä½Žç”µé‡æ—¶å‘é€çš„æ˜¯ CMD_EMERGENCY_STOPï¼ˆç®€åŒ–å®žçŽ°ï¼‰
    // å®žé™…åº”è¯¥å‘é€ NavigationTaskï¼Œè¿™é‡Œå°† EMERGENCY_STOP äºŒæ¬¡åˆ©ç”¨ä¸ºå……ç”µè§¦å‘
    if (cmd.cmd_type() == CMD_EMERGENCY_STOP && battery_ < 20.0) {
        LOG_INFO << "[MockAGV-" << agv_id_ 
                 << "] Low battery detected, interpreting as charge command";
        startMovingToCharger();
    }
}

void MockAgvClient::handleHeartbeat(const Heartbeat& msg) {
    (void)msg;
    LOG_DEBUG << "[MockAGV-" << agv_id_ << "] [RECV] Heartbeat response from server";
    // çœ‹é—¨ç‹—å·²åœ¨ onMessage ä¸­åˆ·æ–°ï¼Œæ— éœ€é¢å¤–å¤„ç†
}

void MockAgvClient::handleNavigationTask(const NavigationTask& task) {
    LOG_INFO << "[MockAGV-" << agv_id_ << "] [RECV] NavigationTask: task_id=" << task.task_id();
    LOG_INFO << "  Target: (" << task.target_node().x() << ", " << task.target_node().y() << ")";
    LOG_INFO << "  Operation: " << OperationType_Name(task.operation());
    LOG_INFO << "  Path points: " << task.global_path_size();
    
    // æ¨¡æ‹Ÿæ‰§è¡Œå¯¼èˆªä»»åŠ¡ï¼š
    // 1. åˆ‡æ¢åˆ° MOVING çŠ¶æ€
    // 2. 3ç§’åŽåˆ‡æ¢å›ž IDLEï¼ˆæ¨¡æ‹Ÿä»»åŠ¡å®Œæˆï¼‰
    setState(MOVING);
    
    // å»¶è¿Ÿä»»åŠ¡ï¼š3ç§’åŽæ¢å¤åˆ° IDLE
    loop_->runAfter(3.0, [this, task_id = task.task_id()]() {
        if (state_ == MOVING) {
            LOG_INFO << "[MockAGV-" << agv_id_ << "] NavigationTask completed: task_id=" << task_id;
            setState(IDLE);
            
            // å¯é€‰ï¼šå‘é€ TaskFeedbackï¼ˆè¿­ä»£ä¸‰åŽç»­å®Œå–„ï¼‰
            // TaskFeedback feedback;
            // feedback.set_agv_id(agv_id_);
            // feedback.set_task_id(task_id);
            // feedback.set_status(TASK_COMPLETED);
            // sendProtobufMessage(MSG_TASK_FEEDBACK, feedback);
        }
    });
}

// ==================== å®šæ—¶ä»»åŠ¡å®žçŽ° ====================

void MockAgvClient::onTelemetryTimer() {
    if (!connected_ || state_ == E_STOP) {
        return;  // æœªè¿žæŽ¥æˆ–æ€¥åœçŠ¶æ€ä¸å‘é€
    }
    sendTelemetry();
}

void MockAgvClient::onHeartbeatTimer() {
    if (!connected_ || state_ == E_STOP) {
        return;  // æœªè¿žæŽ¥æˆ–æ€¥åœçŠ¶æ€ä¸å‘é€
    }
    sendHeartbeat();
}

void MockAgvClient::onBatteryTimer() {
    if (state_ == E_STOP) {
        return;  // æ€¥åœçŠ¶æ€ä¸æ›´æ–°ç”µé‡
    }
    
    double delta = 0.0;
    switch (state_) {
        case IDLE:
            delta = kBatteryDrainIdle;  // -0.5%/s
            break;
        case MOVING:
        case MOVING_TO_CHARGER:
            delta = kBatteryDrainMoving;  // -1.0%/s
            break;
        case CHARGING:
            delta = kBatteryChargeRate;  // +2.0%/s
            break;
        case E_STOP:
            delta = 0.0;  // ä¸å˜
            break;
    }
    
    updateBattery(delta);
    
    // å……ç”µå®Œæˆæ£€æµ‹
    if (state_ == CHARGING && battery_ >= kBatteryMax) {
        onChargingComplete();
    }
}

void MockAgvClient::onWatchdogTimer() {
    if (!connected_) {
        return;  // æœªè¿žæŽ¥ä¸æ£€æŸ¥
    }
    
    Timestamp now = Timestamp::now();
    double elapsed_sec = (now.microSecondsSinceEpoch() - 
                         last_server_msg_time_.microSecondsSinceEpoch()) / 1000000.0;
    
    if (elapsed_sec > watchdog_timeout_sec_ && state_ != E_STOP) {
        LOG_ERROR << "[WATCHDOG] [MockAGV-" << agv_id_ 
                  << "] âš ï¸ [EMERGENCY] Server Lost! "
                  << "(timeout=" << elapsed_sec << "s > " 
                  << watchdog_timeout_sec_ << "s)";
        setState(E_STOP);
    }
}

// ==================== çŠ¶æ€åˆ‡æ¢ä¸Žä¸šåŠ¡é€»è¾‘å®žçŽ° ====================

void MockAgvClient::setState(State new_state) {
    if (state_ == new_state) return;
    
    State old_state = state_;
    state_ = new_state;
    
    LOG_INFO << "[MockAGV-" << agv_id_ << "] State: " 
             << stateToString(old_state) << " -> " << stateToString(new_state);
}

void MockAgvClient::updateBattery(double delta) {
    double old_battery = battery_;
    double new_battery = old_battery + delta;
    
    // é™åˆ¶èŒƒå›´
    if (new_battery < kBatteryMin) new_battery = kBatteryMin;
    if (new_battery > kBatteryMax) new_battery = kBatteryMax;
    
    battery_ = new_battery;
    
    LOG_DEBUG << "[MockAGV-" << agv_id_ << "] Battery: " 
              << old_battery << "% -> " << new_battery << "% (delta=" 
              << delta << "%/s)";
}

void MockAgvClient::refreshServerMessageTime() {
    last_server_msg_time_ = Timestamp::now();
}

void MockAgvClient::onChargingComplete() {
    LOG_INFO << "[MockAGV-" << agv_id_ 
             << "] âš¡ Charging complete (battery=" << battery_ 
             << "%), waiting for RESUME command...";
    // ä¿æŒ CHARGING çŠ¶æ€ï¼Œç­‰å¾…æœåŠ¡å™¨å‘é€ RESUME æŒ‡ä»¤
}

void MockAgvClient::startMovingToCharger() {
    LOG_INFO << "[MockAGV-" << agv_id_ 
             << "] ðŸš— Moving to charger (ETA: " 
             << kMovingToChargerDelaySec << "s)...";
    setState(MOVING_TO_CHARGER);
    
    // æ¨¡æ‹Ÿç§»åŠ¨å»¶è¿Ÿï¼ˆ3ç§’åŽåˆ°è¾¾å……ç”µç‚¹ï¼‰
    loop_->runAfter(kMovingToChargerDelaySec, [this]() {
        if (state_ == MOVING_TO_CHARGER) {
            LOG_INFO << "[MockAGV-" << agv_id_ << "] âš¡ Arrived at charger, start charging";
            setState(CHARGING);
        }
    });
}

// ==================== æ¶ˆæ¯å‘é€å®žçŽ° ====================

void MockAgvClient::sendProtobufMessage(uint16_t msg_type, 
                                        const google::protobuf::Message& message) {
    if (!connected_) {
        return;
    }
    
    // 1. åºåˆ—åŒ– Protobuf
    std::string payload;
    if (!message.SerializeToString(&payload)) {
        LOG_ERROR << "[MockAGV-" << agv_id_ << "] Failed to serialize message";
        return;
    }
    
    // 2. ä½¿ç”¨ LengthHeaderCodec ç¼–ç 
    Buffer buf;
    if (!LengthHeaderCodec::encode(&buf, msg_type, payload)) {
        LOG_ERROR << "[MockAGV-" << agv_id_ << "] Failed to encode message";
        return;
    }
    
    // 3. å‘é€
    conn_->send(buf.retrieveAllAsString());
}

void MockAgvClient::sendTelemetry() {
    AgvTelemetry msg;
    msg.set_agv_id(agv_id_);
    msg.set_timestamp(Timestamp::now().microSecondsSinceEpoch());
    msg.set_x(x_);
    msg.set_y(y_);
    msg.set_theta(theta_);
    msg.set_confidence(0.95);
    msg.set_battery(battery_);
    msg.set_linear_velocity(0.0);
    msg.set_angular_velocity(0.0);
    msg.set_acceleration(0.0);
    msg.set_payload_weight(0.0);
    msg.set_error_code(0);
    msg.set_fork_height(0.0);
    
    sendProtobufMessage(MSG_AGV_TELEMETRY, msg);
    
    LOG_DEBUG << "[MockAGV-" << agv_id_ << "] [SEND] Telemetry: battery=" 
              << battery_ << "%, state=" << stateToString(state_);
}

void MockAgvClient::sendHeartbeat() {
    Heartbeat msg;
    msg.set_agv_id(agv_id_);
    msg.set_timestamp(Timestamp::now().microSecondsSinceEpoch());
    
    sendProtobufMessage(MSG_HEARTBEAT, msg);
    
    LOG_DEBUG << "[MockAGV-" << agv_id_ << "] [SEND] Heartbeat";
}

// ==================== å·¥å…·å‡½æ•° ====================

const char* MockAgvClient::stateToString(State state) {
    switch (state) {
        case IDLE:                return "IDLE";
        case MOVING:              return "MOVING";
        case E_STOP:              return "E_STOP";
        case MOVING_TO_CHARGER:   return "MOVING_TO_CHARGER";
        case CHARGING:            return "CHARGING";
        default:                  return "UNKNOWN";
    }
}
